### ISSUE TITLE
Option to specify separator for "Copy Relative Path" action

### ISSUE THREAD

**Issue by IllusionMH** [microsoft/vscode#56279]
**Workaround (from @bpasero)**
You can use this extension: https://marketplace.visualstudio.com/items?itemName=rssowl.copy-relative-path-posix

**Original Issue:**

Looked for other mentions but haven't found related requests or discussion.

I'm often using Copy Relative Path feature to compose import strings or share with other developers on Mac/Linux. Therefore, new Copy relative path feature that has native path separators on Windows doesn't help in these cases (in comparison with Copy path).

Please add option to change path separators used in Copy Relative Path action, so I will be able to set `/` as path separator and have same behavior across all platforms.

**Comment by isidorn** [comment]
Both copy path and copy relative path respect the platform they are on
I can leave this open as a feature request. Add a setting just for this sounds like a bad idea to me

fyi @bpasero

**Comment by IllusionMH** [comment]
@isidorn proposed option will open more use cases for Windows users (other have them "for free").
JS/TS imports expect path with forward slashes independently from OS, as well as paths for images/resources used in HTML/CSS. Ability to copy path with expected separator will remove pain from manually changing them (which should be done with Copy Path).

I've proposed to make it as option, because it is already released in this way and separator change might be considered breaking change. Moreover this will provide same behavior as in WebStorm: relative path - always forward slashes, (full) path - according to OS.

**Comment by yumeyao** [comment]
#6308

There was a closed issue with similar request, but was rejected by the arrogant developer.

**Comment by esmirnov** [comment]
There is extension for that https://marketplace.visualstudio.com/items?itemName=alexdima.copy-relative-path

**Comment by esmirnov** [comment]
@IllusionMH agree, it's not perfect. I would suggest a new setting "preferedPathSepartor" with possible values "Native", "Forward", "Backward".

**Comment by dwelle** [comment]
I'd so love an option to set POSIX (forward slash) separators everywhere in vscode on windows. I hate manually replacing\escaping slashes everywhere. I'm not using `cmd.exe`, windows explorer accepts POSIX paths, so why do we need backslashes at all?

Please add an option to display POSIX all the way.

**Comment by efortes** [comment]
I often use the vscode Windows Subsystem for Linux (WSL) terminal to navigate to a file by right clicking a file in VS code and hitting the Copy path options. This will deliver me a windows path ( C:\Users\. ).
As i'm using WSL I want the possibility to copy an Unix path. The path would then be: /c/Users. ( https://github.com/Microsoft/vscode/issues/63633 )

**Comment by mandrie** [comment]
+1 for making the slash direction optional to the user. I do a lot of work with Markdown file on Windows in VSCode for [Azure Devops Wiki](https://docs.microsoft.com/en-us/azure/devops/project/wiki/markdown-guidance?view=vsts#source-control-relative-links) and relative path file links need to use forward slashes and not backslashes, so I am constantly using the **"Copy Relative Path"** and replacing the backslashes

**Comment by victor-ruben** [comment]
This extension does that: https://github.com/CoenraadS/Copy-WSL-Path

**Comment by IllusionMH** [comment]
@victor-ruben unfortunately this extension doesn't help. It will produce full path like `/mnt/c/long/path/to/project/src/file/project.js` ([see](https://github.com/CoenraadS/Copy-WSL-Path/blob/1201e598b3ae185891dd12f4a99beaccaca1fe9f/src/extension.ts#L10)) but in this issue we are asking for path relative for workspace. So it will be simple `src/file/project.js`.

**Comment by dwelle** [comment]
@IllusionMH there's https://marketplace.visualstudio.com/items?itemName=dkaba.better-file-path

**Comment by gomesalexandre** [comment]
What's the status on this ? It would be really nice to have options for relative file path copy, same way as you do for the window title for example. Given this is already implemented there, it should be quite trivial

**Comment by gomesalexandre** [comment]
I would like to contribute it if possible, just not sure where to get started

**Comment by libasoles** [comment]
+1 for this. Makes total sense for me.

**Comment by victor-ruben** [comment]
> 
> 
> I would like to contribute it if possible, just not sure where to get started

Perhaps a PR to this extension https://github.com/CoenraadS/Copy-WSL-Path

**Comment by Juxn3** [comment]
+1 for this

**Comment by kyletheninja** [comment]
> I would suggest a new setting "preferedPathSepartor" with possible values "Native", "Forward", "Backward".

esmirnov proposes the perfect solution in my eyes. I recently started using VS code for angularJS development and the copy relative path function would be fantastic for pasting template links into directives and js links into index.html, but backslash is an escape character in JS. It's super annoying to have this great and useful feature have such a debilitating limitation, which is how I would describe the situation as pertains to use in JS.

**Comment by maplion** [comment]
Can there just be an added context menu option that one can turn on, such as "Copy Relative Path with /"?  I would find this very useful as even being a Windows user, I use the forward slash path more often than the backslash one.

**Comment by bobbogomip** [comment]
I use _vscode_ _because_ it is portable.

 - Native Windows APIs accept forward slashes or back slashes.
 - Relative paths in all my code therefore use forward slashes. Some languages mandate this.
 - When on Windows I use _bash_ as my shell (_vscode_ lets you do this).
 - Just about the only place where you must use backslashes is in _cmd.exe_ as far as I can tell.

I _never_ use _cmd_.

Please _please_ can Windows _vscode_ support some way of getting paths with forward slashes?

**Comment by xmadsen** [comment]
An obvious feature, should definitely exist for WSL users who want to work with Linux users and not look like idiots using Windows slashes :+1:

**Comment by ElvenSpellmaker** [comment]
I'd like to chime in and say I use Cygwin through Mintty and the backslashes are really killing my productivity due to them being forward slashes, and not being able to configure a full path output. =|

**Comment by pianeer** [comment]
This is a necessary feature and huge productivity booster for those of us who use Linux shells in Windows. Could we please add this as a configuration option with the default set to the native but at least have the ability to override it?

**Comment by EletronicElephant** [comment]
I think it's OK to leave the option to the user. It's really needed.

**Comment by vadimmos** [comment]
The ability to choose a separator is something that is so lacking.

**Comment by hornta** [comment]
Or don't specify an option and always return a path with "/"

**Comment by 3nuc** [comment]
Especially now when WSL is a thing this is important. Even without WSL there are shells that use the Linux format - git bash for example. Switching to stock cmd just because I need to paste a VS Code path is a pain.

**Comment by bpasero** [comment]
I just released an extension that should take care of this (as I still think this should be something an extension provides and not be a setting): https://marketplace.visualstudio.com/items?itemName=rssowl.copy-relative-path-posix

![image](https://user-images.githubusercontent.com/900690/67633511-ef0c3d00-f8b0-11e9-818b-ece1d7137373.png)


**Features:**
* supports multi-select in explorer
* supports editor title context menu
* will not add the name of the workspace folder if in a multi-folder setup
* has only minimal dependencies (e.g. uses the official API to copy to clipboard)

If there are no objections and people think the extension is OK enough, I would suggest to close this issue.

**Comment by lalomartins** [comment]
Thanks for the extension, but I think there's quite enough argument above that this should be built-in. Whether it's a setting, or a separate menu entry like in the extension, or simply always copy the posix slash (which works everywhere except `cmd` as was pointed out earlier, I don't know, but it's not something to be left to an extension IMO.

**Comment by saschanaz** [comment]
I think it should be reversed, the default should be standard POSIX and the cmd compatibility should be an extension.

**Comment by bpasero** [comment]
I pushed an update for the extension so that it can also be used via keybinding or from the command palette. In that case of usage it will copy the relative path of the active editor.

**Comment by maplion** [comment]
@bpasero Thanks for this, I'm going to use it until there is an alternative.

**Comment by Orbling** [comment]
This really should be a standard feature, guess the developers don't actually develop on Windows platforms much.

Few developers are going to use cmd.exe as a terminal, it'll usually be bash, often as not people will be developing for non-windows environments.  Having the copy path options forcibly use a type of slash that's wrong in the majority of usage cases is daft and makes the feature next to useless.  It really does need to be settings-based or have two completely separate options.  May well want to be per language, as desktop applications in an MS flavour might still want the backslash.

**Comment by Rob4226** [comment]
Extensions are nice but please add it as a native setting so it can be set on a per-folder or per-workspace basis in settings.json.  This is especially needed for developing Angular apps.  It's a reasonable request with good support and no real downsides.  It can default to false to keep current behavior but something like the following would be very helpful. Thank you.
"editor.copyRelativePathForwardSlashes": true

**Comment by dacioromero** [comment]
>  It can default to false to keep current behavior but something like the following would be very helpful. Thank you.
> "editor.copyRelativePathForwardSlashes": true

I'm thinking that `"editor.copyRelativePathSlashType"` would be a better option that could have `"forward"`, `"back`", or `"os"` (default). That way if someone wanted backslashes over forward for whatever reason they have the option.

**Comment by eliveikis** [comment]
Another upvote. Especially with Windows Subsystem for Linux 2 incoming, more and more people are using Windows combined with linux workflows. I personally use git bash or an Ubuntu terminal. For all day-to-day windows work.

In addition, many development projects use forward slashes for path definitions that are passed to configs, command lines or API's. Not having this as a VSCode option feels very backwards and is a common pain point.

Even windows explorer accepts forward slashes in the address bar.

**Comment by bpasero** [comment]
Quick poll: what if there was simply a setting that puts any path on Windows with slash instead of backslash? Changing this just for the "Copy Path" action seems a bit inconsistent. Would that be OK for people?

**Comment by lalomartins** [comment]
For me that's definitely desirable (more than OK).  Again, the only thing I'm aware of which still *requires* backslash is `cmd`, everything else may *accept* it but also accepts slash, so as I see it there's no real reason to be using backslash.

(And if you're a developer and you're using `cmd`, please do yourself a favor and take a couple of days to learn PowerShell, it's 2020 already.)

**Comment by dwelle** [comment]
@bpasero I'm pretty sure that's actually what everyone wants, but they settled for "path". Bring it on! :)

**Comment by IllusionMH** [comment]
@bpasero overall it will be even more preferable especially if implemented for Compact folders in Explorer and  search results and open editors description text (relative path).

Not so optimal for full paths in hovers (tabs, open editors, explorer), but not a problem.

But problem will be reversed to "Copy path" which copies full path and OS separators make more sense here.

TL;DR; I'm all in for this option, but proper OS path for Copy path makes more sense

**Comment by bpasero** [comment]
Maybe "Copy Path" could be the only special case where we would use backslash even if this new setting is configured to use slashes. I agree that an absolute path when copied should use backslash on Windows.

**Comment by sergey-beloglazov** [comment]
I need 2 options for "Copy relative path":
1) for Windows, where I work with IDE
2) for Unix, which is the target for my PHP files
The solution with native function with OS separator + plugin suites for me.
If there will be an option for separator, it doesn't solve the problem dealing with 2 OS file systems :)

**Comment by dwelle** [comment]
@sergey-beloglazov if you want to retain platform-specific conventions then simply don't change the (future) setting. Unix is already POSIX, windows will keep backslashes.

This issue is about support of overriding the convention for all platforms.

**Comment by LeanHer94** [comment]
Also would be great if you fix class discovery. That way we wouldn't need to go copy (type in this case) the path to the file that we wanna import in our .ts

**Comment by djmisterjon** [comment]
yes also need it.
Maybe the `alt` + `click` on the command can be a good idea!, or `shift` + `rightClick` to make appear secret commands like in Window explorer?
A option to toggle between path mode can also a great idea. 
We lost precious time to convert path slash `\`  `/` and a plugin to just do this ! no thanks

**Comment by wonkoRT** [comment]
I'd like this, additionally with an option to copy escaped paths in source code and unescape them while copying and vice versa. Also it would be great if pasting into a source would recognize paths and auto escape them and also convert forward to backslash paths prefixed with \\?\ (escaped \\\\?\\) at least when hitting the max path length limit.

**Comment by IllusionMH** [comment]
@wonkoRT escaping/unescaping should be separate request because need to escape/unescape depends too much on context of from/to operation (config vs string vs import location, language etc.).

This one about removing difference between how paths(mostly relative) are handled in windows vs *nix for display, copy etc.

**Comment by infacto** [comment]
I would also prefer an option for that. One use case: Right click, copy the relative path and insert to the terminal for e.g. run tests. But the backslashes are not compatible in this case. So I have to replace the slashes every time.

For the implementation of this option in vscode this may help:

```ts
/** Replaces all back or forward slashes to sepcific slash type. (Default: `/`) */
const normalizePath = (path: string, slash = '/'): string => path.replace(/[/\\]/g, slash);
```

**Comment by Rah1x** [comment]
sooo? when will this happen?

**Comment by rickvian** [comment]
omg it still open, i see there are lots of Upvotes in the first issue opener, does that really help, or we should keep bumping with new comment in issue to get attention to this discussion? its been 2 year since this issue opened

**Comment by bpasero** [comment]
Today I went back and forth whether a setting for this should apply to all labels displayed or be specific to the copy relative path action and came to the conclusion that it should be specific to "Copy Relative Path" as originally requested because that is the main issue at hand here: you want to copy a relative path to use it in a source code file that asks for slashes. 

With https://github.com/microsoft/vscode/pull/101197 being merged now, there is a new setting `explorer.copyRelativePathSeparator` with the options `/` and `\`. 

**Verification:**
* without configuring this setting, copying a relative or absolute path should use slash on macOS/Linux and backslash on windows
* when configuring this setting, verify the configured path separator is used over the standard OS one but only for "Copy Relative Path"

This change will be available for feedback once we have released `1.58` this week and insiders starts again (I would think this week Friday).
